Make volume binder resilient to races

cofyc opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one, leave it on its own line:
>
> /kind api-change
/kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

Make volume binder resilient to races between main schedule loop and async binding operation.

**Which issue(s) this PR fixes**:
<!-- 
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #71928 #72013 #56236

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE".
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Make volume binder resilient to races between main schedule loop and async binding operation 
```

/sig storage
/sig scheduling

cofyc commented:
/assign @msau42

msau42 commented:
Hm another idea I had is to treat static binding and pending provisioning similarly in the predicate.  Instead of failing the predicate if selectedNode is set, can we instead make the predicate return true only for the selectedNode, and false for all other nodes?  AssumePodVolumes checks for if the PVCs are fully bound (which they are not) to determine whether or not to call BindPodVolumes.

This would be a bigger change, but I think would make our logic more resilient to these races.

cofyc commented:
That makes sense and seems better! Always try to bind pod volumes if they are not fully bound.

cofyc commented:
> Instead of failing the predicate if selectedNode is set, can we instead make the predicate return true only for the selectedNode, and false for all other nodes? AssumePodVolumes checks for if the PVCs are fully bound (which they are not) to determine whether or not to call BindPodVolumes.

The important thing in this design is how to make API calls and check on new bindings made by new schedule loop. IMO there are two methods:

- Method 1: when new bindings are found, make API calls for new bindings and check on new bindings
- Method 2: fail with error when new bindings are found, let scheduler fire new binding operation next run
  - Because pod is not marked unscheduable, pod will be added into active queue on failure

With method 1, schedule loop which made new bindings will continue to retry until binding operation finished. With method 2, two instances of pod will merge into one in active queue, this reduces unnecessary scheduling retries.

I've update PR to use method 2. Please take a look at again.

cofyc commented:
## Scenarios

#### PVC selectedNode is rejected by provisioner

xref: https://github.com/kubernetes/kubernetes/issues/71928

BindPodVolumes will be retired by following cases:

- No pod is in scheduling. New PVC object from apiserver is added into pvc cache. `BindPodVolumes` find `selectedNode` does not exist, fail with error on which scheduler will retry.
- A pod is in scheduling. It will assume on new PVC object and update binding cache. `BindPodVolumes` will find binding cache is out of date, fail with an error on which scheduler will retry

cofyc commented on pkg/controller/volume/persistentvolume/scheduler_binder.go:
```diff
@@ -141,6 +141,27 @@ func (b *volumeBinder) GetBindingsCache() PodBindingCache {
+		if len(provisionedClaims) == 0 {
+			provisionedClaims = nil
+		}
+		// TODO merge into one atomic function
```
In current implementation, IMO there is no need to separate `bindings` and `provisionedPVCs` in pod binding cache. It's better to get/update/delete `bindings` and `provisionedPVCs` together.

cofyc commented:
/retest

cofyc commented on pkg/controller/volume/persistentvolume/scheduler_binder.go:
```diff
@@ -382,8 +423,13 @@ func (b *volumeBinder) checkBindings(pod *v1.Pod, bindings []*bindingInfo, claim
 	for _, binding := range bindings {
-		// Check for any conditions that might require scheduling retry
+		// Check if pv is out of date
+		if err := b.pvCache.IsOutOfDate(binding.pv); err != nil {
```
This causes lots of unnecessary scheduling retires which increase volume binding latency. TestVolumeBindingDynamicStressFast/TestVolumeBindingDynamicStressSlow have high chance to fail. I'm working on a more fine-grain version.

cofyc replied:
Done

cofyc commented on pkg/controller/volume/persistentvolume/scheduler_binder.go:
```diff
@@ -397,18 +464,26 @@ func (b *volumeBinder) checkBindings(pod *v1.Pod, bindings []*bindingInfo, claim
 		}
-
-		// TODO; what if pvc is bound to the wrong pv? It means our assume cache should be reverted.
-		// Or will pv controller cleanup the pv.ClaimRef?
```
https://github.com/kubernetes/kubernetes/blob/197fc67693c2391dcbc652fc185ba85b5ef82a8e/pkg/controller/volume/persistentvolume/pv_controller.go#L652-L699

If PVC is bound to wrong PV, PV will be unbound. Assumed cache will be reverted.
I've added a test for dynamic provisioned PV by the way.

msau42 replied:
There is a PR here that will detect if PV has invalid node affinity and will fail binding: https://github.com/kubernetes/kubernetes/pull/70026

cofyc commented on pkg/controller/volume/persistentvolume/scheduler_binder_test.go:
```diff
@@ -1275,60 +1404,70 @@ func TestCheckBindings(t *testing.T) {
+			binding:    makeBinding(unboundPVC, pvNode1aBound),
+			apiPV:      pvNode1aBound,
+			apiPVC:     boundPVCNode1a,
+			shouldFail: true,
```
If PV is already bound, BindPodVolumes actually will fail in bindAPIUpdate because object version conflict.

cofyc replied:
Sorry, I misunderstood 'already-bound' test.  There are more cases:

Case 1: Before FindPodVolumes, PV is bound to PVC, PVC is not fully bound to PV

IIUC, this is original test case. BindPodVolumes will succeed after PVC is fully bound to PV by PV controller. I implement this behavior by using asyncFunc.

Case 2: Before FindPodVolumes, PV is not bound to PVC, PVC is not fully bound to PV

If PV is bound to PVC by PV controller before BindPodVolumes, BindPodVolumes will fail because API conflict.

I've added a test `bound-by-pv-controller-before-bind` for this.

Case 3: Before FindPodVolumes, PV is bound to PVC, PVC is fully bound to PV

BindPodVolumes will not be called. So we do not need test this.

msau42 commented on pkg/controller/volume/persistentvolume/scheduler_binder.go:
```diff
@@ -382,13 +436,33 @@ func (b *volumeBinder) checkBindings(pod *v1.Pod, bindings []*bindingInfo, claim
+
 	for _, binding := range bindings {
-		// Check for any conditions that might require scheduling retry
+		// Check against latest API object
```
Can you explain why we need to check the API object directly and not wait for the informer update to the pv cache?

cofyc replied:
Because PV cache may be overridden. Binding operation is async, scheduling in main loop may Find/Assume new cache on new API objects, then binding operation will don't know latest API objects. This is root cause of volume binder races (e.g. https://github.com/kubernetes/kubernetes/issues/71928). 

e.g. 

- Volume binder find/assume PV/PVC cache (cache: prebound, api: not bound)
- Volume binder make API call to bind PV/PVC and check PV & PVC is bound or not (cache: prebound, api: prebound)
- PV/PVC is rejected (cache: not bound, api: not bound)
- Before binding operation check PV/PVC cache, Volume binder find/assume PV/PVC cache again (cache: prebound, api: not bound)
- Binding operation finds PV/PVC cache are still the same, but actually binding operation should be restarted to bind PV/PVC again

In current scheduler, we can simply do not share PV/PVC cache in two goroutines to avoid races. IMO, PV/PVC cache are used for scheduling, not for binding operation check which should depends on latest state in API server. Because PV/PVC are bound or not and pod should be rescheduled or not depend on state in API server which are updated by PV controller or provisioner. 

Though PV/PVC cache can be used if they will not be modified by another one if we prevent scheduler from scheduling pod when pod is in binding , but it is still unnecessary.

msau42 replied:
Ok that makes sense. Can you update the comment here with a little more detail, saying that because the PV/PVC can be assumed again, we need to fail the binding operation so it can be retried.

And actually, assume pod will fail because binding is in progress.  After this binding operation fails, then what causes the pod to retry scheduling again?

cofyc replied:
Yes, if pod is not marked unschedulable (status=PodReasonUnschedulable) wrongly (fixed in this PR), [other status updates will re-add pod into active queue](https://github.com/kubernetes/kubernetes/blob/5b6e20d0e057c0060a185673adc75af44d1a4566/pkg/scheduler/internal/queue/scheduling_queue.go#L367). This will retry scheduling again.

cofyc replied:
assume pod [fails with non-PodReasonUnschedulable error](https://github.com/kubernetes/kubernetes/blob/5b6e20d0e057c0060a185673adc75af44d1a4566/pkg/scheduler/scheduler.go#L410-L416) which will not put pod into unscheduable pod.

cofyc replied:
Comments are updated in 80d8815f64fb1452d8c62c2513707b90401d2b10.

msau42 commented on pkg/controller/volume/persistentvolume/scheduler_binder.go:
```diff
@@ -382,13 +436,33 @@ func (b *volumeBinder) checkBindings(pod *v1.Pod, bindings []*bindingInfo, claim
 	for _, binding := range bindings {
-		// Check for any conditions that might require scheduling retry
+		// Check against latest API object
+		obj, err := b.pvCache.GetAPIObj(binding.pv.Name)
```
Maybe add a new method `GetApiPV` [here](https://github.com/kubernetes/kubernetes/blob/cdeb9205230ac6e5ea772ab8c562738d7c42bc3e/pkg/controller/volume/persistentvolume/scheduler_assume_cache.go#L341) to make the logic here easier to read.

cofyc replied:
Done in 80d8815f64fb1452d8c62c2513707b90401d2b10.

msau42 commented on pkg/controller/volume/persistentvolume/scheduler_binder.go:
```diff
@@ -300,10 +345,8 @@ func (b *volumeBinder) BindPodVolumes(assumedPod *v1.Pod) (err error) {
-		bindings = b.podBindingCache.GetBindings(assumedPod, assumedPod.Spec.NodeName)
-		claimsToProvision = b.podBindingCache.GetProvisionedPVCs(assumedPod, assumedPod.Spec.NodeName)
-		return b.checkBindings(assumedPod, bindings, claimsToProvision)
+		b, err := b.checkBindings(assumedPod, bindings, claimsToProvision)
```
Is the pod deletion case still handled?

cofyc replied:
Oh, I missed this, binding operation should be cancelled when pod is deleted or assigned a node by users or other scheduler.

Done in 7140fda91d1a239862b3347977fafe27efab1994.

cofyc commented:
/test pull-kubernetes-verify

msau42 commented on pkg/controller/volume/persistentvolume/scheduler_binder.go:
```diff
@@ -442,6 +442,15 @@ func (b *volumeBinder) checkBindings(pod *v1.Pod, bindings []*bindingInfo, claim
+	// to check PV/PVC bindings any more.
+	// We check pod binding cache here which will be cleared when pod is
+	// removed from scheduling queue.
+	if b.podBindingCache.GetDecisions(pod) == nil {
```
Can you add a unit test for this, if one doesn't already exist?

cofyc replied:
sure, it has some conflicts with #70026, I will rebase and squash after #70026 is merged.

cofyc replied:
There is [an existing test `pod-deleted-after-time`](https://github.com/kubernetes/kubernetes/pull/72045/files#diff-7e773cadec5750de298ed8fa194edda3L1357). In old PR, BindPodVolume failed with timeout error, so it passed too. Perhaps it's better to check error too. I can address this in a separate PR.

msau42 reviewed (commented):
lgtm, just one minor comment and this can be squashed a little

Commit:
Make volume binder resilient to races

- FindPodVolumes do not error if PVC is assumed with selected node
- BindPodVolumes check against API objects

Commit:
Make volume binder resilient to races: unit tests

Commit:
Make volume binder resilient to races: scheduler change

There is no need to clear stale pod binding cache in scheduling, because
it will be recreated at beginning of each schedule loop, and will be
cleared when pod is removed from scheduling queue.

Commit:
If provisioning PVC's PV is not found, check next time.

cofyc commented:
@msau42 PR is rebased and squashed into logical commits:

- 13d87fbff8ce4030bce5209caedbb2e2d55e13bf: Make volume binder resilient to races 
- 4f8772e74bc1f1d20e4ca4de467d882234c141b1: Unit tests only
- d0e748a47f2e5da843ce8d8e19c35f0b9be9689f: Scheduler change
- f7aca60aaf77b6691d1e219b985479b46e7454f5: If provisioning PVC's PV is not found, check next time

I made an additional modification to original PR, tolerate provisioning PVC's PV not found error, see https://github.com/kubernetes/kubernetes/pull/72045/commits/f7aca60aaf77b6691d1e219b985479b46e7454f5. This avoids unnecessary "failed to get pv" errors and scheduler retires.

msau42 commented on pkg/controller/volume/persistentvolume/scheduler_binder.go:
```diff
@@ -518,7 +518,15 @@ func (b *volumeBinder) checkBindings(pod *v1.Pod, bindings []*bindingInfo, claim
 			if err != nil {
-				return false, fmt.Errorf("failed to get pv %q from cache: %v", pvc.Spec.VolumeName, err)
+				if _, ok := err.(*errNotFound); ok {
+					// We tolerate NotFound error here, because PV is possibly
```
unit test for this condition?

msau42 commented on pkg/controller/volume/persistentvolume/scheduler_binder_test.go:
```diff
@@ -1302,63 +1431,110 @@ func TestCheckBindings(t *testing.T) {
+			apiPVC:     boundPVCNode1a,
+			shouldFail: true, // bindAPIUpdate will fail because API conflict
+		},
+		"binding-static-pv-succeeds-after-time": {
```
What's the difference between this and "binding-succeeds-after-time"? That the PV object is not bound?

cofyc replied:
they are same... my mistake, i'm going to remove one of them.

msau42 commented on pkg/controller/volume/persistentvolume/scheduler_binder_test.go:
```diff
@@ -1302,63 +1431,110 @@ func TestCheckBindings(t *testing.T) {
+					t.Errorf("failed to create PV %q: %v", dynamicPV.Name, err)
+					return
+				}
+				newPVC.Spec.VolumeName = dynamicPV.Name
```
does this need to be a copy of the pvc object?

cofyc replied:
No, `newPVC` is returned by `client.CoreV1().PersistentVolumeClaims(pvc.Namespace).Get` which retrieves objects from apiserver. I've updated volumeReactor to return a copy of stored object.

cofyc replied:
For dynamic provisioning, PVC was updated in bindAPIUpdate, this is why I need to fetch PVC from apiserver again.

msau42 commented on pkg/controller/volume/persistentvolume/scheduler_binder_test.go:
```diff
@@ -1376,23 +1552,23 @@ func TestBindPodVolumes(t *testing.T) {
+			delayFunc: func(t *testing.T, testEnv *testEnv, pod *v1.Pod, pvs []*v1.PersistentVolume, pvcs []*v1.PersistentVolumeClaim) {
+				pvc := pvcs[0]
 				// Delete PVC
 				// Update pvc cache, fake client doesn't invoke informers
```
can you update the rest of these tests to not directly update the internal cache since it's not needed anymore?

msau42 commented on pkg/controller/volume/persistentvolume/scheduler_binder_test.go:
```diff
@@ -1445,38 +1621,54 @@ func TestBindPodVolumes(t *testing.T) {
+		}
+
+		// Before Execute
+		if scenario.apiPV != nil {
```
What is the purpose of having a separate call to update the api objects instead of passing them through initVolumes/initClaims?

cofyc commented on pkg/scheduler/factory/factory.go:
```diff
@@ -1124,21 +1123,11 @@ func (c *configFactory) MakeDefaultErrorFunc(backoff *util.PodBackoff, podQueue
-
-					if c.volumeBinder != nil {
-						// Volume binder only wants to keep unassigned pods
-						c.volumeBinder.DeletePodBindings(origPod)
```
In `deletePodFromSchedulingQueue`:

https://github.com/kubernetes/kubernetes/blob/55d8f709c8b214f72c000076093e08344dc3ebce/pkg/scheduler/factory/factory.go#L580-L603

cofyc replied:
Have revisited `DeletePodBindings` part again. I found `deletePodFromSchedulingQueue` is called only [when a unscheduled pod is deleted](https://github.com/kubernetes/kubernetes/blob/55d8f709c8b214f72c000076093e08344dc3ebce/pkg/scheduler/factory/factory.go#L314-L338).

After this PR, we do not need to clear pod binding cache on error because we recreate at beginning of scheduler loop, but we had never cleared cache on success.

IMO, pod binding cache at least should be cleared when pod is not needed to be scheduled (e.g. assigned a node on success or deleted from cluster).

Don't want to make this PR too complex, I created a issue to track this: https://github.com/kubernetes/kubernetes/issues/72763

cofyc replied:
Sorry, I misunderstood [this part code](https://github.com/kubernetes/kubernetes/blob/55d8f709c8b214f72c000076093e08344dc3ebce/pkg/scheduler/factory/factory.go#L314-L338), `deletePodFromSchedulingQueue` will be called on success, because with FilteringResourceEventHandler, object stops passing the filter (assigned a pod or not here) after an update is considered a delete. So all successful pods will go through `deletePodFromSchedulingQueue`.

Talked with @msau42 offline, we'll add integration tests for verify no memory leaks on pod binding cache.

cofyc commented on pkg/controller/volume/persistentvolume/framework_test.go:
```diff
@@ -326,7 +326,7 @@ func (r *volumeReactor) React(action core.Action) (handled bool, ret runtime.Obj
 		name := action.(core.DeleteAction).GetName()
 		klog.V(4).Infof("deleted claim %s", name)
-		obj, found := r.volumes[name]
+		obj, found := r.claims[name]
```
This is weird bug... perhaps old tests didn't reach here

cofyc commented:
/test pull-kubernetes-bazel-test

msau42 commented:
lgtm! thanks for digging into these tricky race conditions

cofyc commented:
@msau42 Squashed into logical commits (no content change):

- 13d87fbff8ce4030bce5209caedbb2e2d55e13bf: Make volume binder resilient to races 
- 8b94b9625bb7bbf0a1d9f8c03f715a4aa76ded1a: Unit tests only
- cfc8ef51d1f57f6dab061fa26fbc14db5dde14e4: Scheduler change
- 1a62f53d3fc915a1c99156c8de049fedd2219a41: If provisioning PVC's PV is not found, check next time

msau42 commented:
/lgtm
/approve

cofyc commented:
cc @k82cn @bsalamat 
For approval on scheduler change

cofyc commented:
/hold
for checking pod binding cache issues

cofyc commented:
/hold cancel

xref: https://github.com/kubernetes/kubernetes/pull/72045#discussion_r246655042

cofyc commented:
/priority important-soon

bsalamat reviewed (approved):
/approve

I do not know the volume binder logic very well, but based on your explanation that the volume binder cache needs to be kept only for unassigned pods, the changes in the scheduler code look good to me.

msau42 commented:
With a quick glance, I don't think so. It seems like only containerd jobs with CSI hostpath or intree nfs drivers have spiked.  Also some of the nfs tests that are failing don't use PVs. I more suspect an issue with containerd

liggitt commented:
good observation. opened https://github.com/kubernetes/kubernetes/issues/72863

bsalamat commented:
We are also seeing an increase in scheduling latency right after this PR. The scheduler predicate evaluation latency has increase by about 50%.

cofyc commented:
I guess it is because new FindPodVolumes in this PR needs to clear old pod binding cache, and to achieve this I update pod cache even if there is no PVCs to bind/provision. There are some optimizations we can do. Sorry about increased scheduling latency. I'm working on a fix.

cofyc commented:
Thanks!
